simulators/ethereum/engine/main.go

```diff
@@ -0,0 +1,177 @@
+	About string
+	Run   func(*TestEnv)
+}
+
+var tests = []testSpec{
+
+	// Engine API Negative Test Cases
+	{Name: "Engine API Proof of Work", Run: engineAPIPoWTests},
+	{Name: "Unknown HeadBlockHash", Run: unknownHeadBlockHash},
+	{Name: "Unknown SafeBlockHash", Run: unknownSafeBlockHash},
+	{Name: "Unknown FinalizedBlockHash", Run: unknownFinalizedBlockHash},
+
+	// Eth RPC Status on ForkchoiceUpdated Events
+	{Name: "Latest Block after ExecutePayload", Run: blockStatusExecPayload},
+	{Name: "Latest Block after New HeadBlock", Run: blockStatusHeadBlock},
+	{Name: "Latest Block after New SafeBlock", Run: blockStatusSafeBlock},
+	{Name: "Latest Block after New FinalizedBlock", Run: blockStatusFinalizedBlock},
+	{Name: "Latest Block after Reorg", Run: blockStatusReorg},
+
+	// Suggested Fee Recipient in Payload creation
+	{Name: "Suggested Fee Recipient Test", Run: suggestedFeeRecipient},
+
+	// Random opcode tests
+	{Name: "Random Opcode Transactions", Run: randomOpcodeTx},
+}
+
+func main() {
+	suite := hivesim.Suite{
+		Name: "engine",
+		Description: `
+Test Engine API tests using CL mocker to inject commands into clients after they 
+have reached the Terminal Total Difficulty.`[1:],
+	}
+	suite.Add(&hivesim.ClientTestSpec{
+		Name:        "client launch",
+		Description: `This test launches the client and collects its logs.`,
+		Parameters:  clientEnv,
+		Files:       files,
+		Run:         runSourceTest,
+	})
+	hivesim.MustRunSuite(hivesim.New(), suite)
+}
+
+// First launch a client against other clients will sync and run their respective tests.
+func runSourceTest(t *hivesim.T, c *hivesim.Client) {
+	clMocker = NewCLMocker()
+	vault = newVault()
+
+	ec := NewEngineClient(t, c)
+	clMocker.AddEngineClient(ec)
+
+	enode, err := c.EnodeURL()
+	if err != nil {
+		t.Fatal("can't get node peer-to-peer endpoint:", enode)
+	}
+	newParams := clientEnv.Set("HIVE_BOOTNODE", enode)
+
+	t.RunAllClients(hivesim.ClientTestSpec{
+		Name:        fmt.Sprintf("%s", c.Type),
+		Description: "test",
+		Parameters:  newParams,
+		Files:       files,
+		Run:         runAllTests,
+	})
+	clMocker.stopPoSBlockProduction()
+}
+
+// Run all tests against a second running client which must sync at some point to the PoS chain.
+func runAllTests(t *hivesim.T, c *hivesim.Client) {
+	// Create an Engine API wrapper for this client
+	ec := NewEngineClient(t, c)
+	defer ec.Close()
+	// Add this client to CLMocker
+	clMocker.AddEngineClient(ec)
+	defer clMocker.RemoveEngineClient(ec)
+	s := newSemaphore(16)
+	var cancelFP *context.CancelFunc
+	var m sync.Mutex
+	var st bool
+	go func() {
+		WaitClientForPoSSync(t, c, &cancelFP, &m, &st)
+	}()
+	for _, test := range tests {
+		test := test
+		s.get()
+		go func() {
+			defer s.put()
+			t.Run(hivesim.TestSpec{
+				Name:        fmt.Sprintf("%s (%s)", test.Name, c.Type),
+				Description: test.About,
+				Run: func(t *hivesim.T) {
+					runTest(test.Name, t, c, ec, vault, clMocker, &m, &st, test.Run)
+				},
+			})
+		}()
+	}
```

Comment:
This is pretty unusual way of doing this, what's more common is to use a `sync.WaitGroup`. 

```suggestion
	var wg sync.WaitGroup
	for _, test := range tests {
		wg.Add(l1)
		test := test
		go func() {
			defer wg.Done()
			t.Run(hivesim.TestSpec{
				Name:        fmt.Sprintf("%s (%s)", test.Name, c.Type),
				Description: test.About,
				Run: func(t *hivesim.T) {
					runTest(test.Name, t, c, ec, vault, clMocker, &m, &st, test.Run)
				},
			})
		}()
	}
```
and later on, `wg.Wait()` instead of `s.drain()`.

Reply:
Worked perfectly, thanks!